storagemarket/sealingpipeline/sectorslist.go

```diff
@@ -0,0 +1,240 @@
+	Active        bool
+	DealWeight    big.Int
+	VerifiedPower big.Int
+	Expiration    abi.ChainEpoch
+	OnChain       bool
+}
+
+type onUpdateCb func(error)
+
+type SectorsList struct {
+	listLk      sync.Mutex
+	list        []SectorState
+	initOnce    sync.Once
+	initialized chan struct{}
+	updatedAt   time.Time
+
+	syncUpdate chan onUpdateCb
+
+	fullNode v1api.FullNode
+	spApi    API
+}
+
+func NewSectorsList(fullNode v1api.FullNode, spApi API) *SectorsList {
+	return &SectorsList{
+		initialized: make(chan struct{}),
+		syncUpdate:  make(chan onUpdateCb, 256),
+		fullNode:    fullNode,
+		spApi:       spApi,
+	}
+}
+
+func (l *SectorsList) Run(ctx context.Context) {
+	timer := time.NewTicker(10 * time.Minute)
+	defer timer.Stop()
+
+	update := func() error {
+		err := l.update(ctx)
+		if err != nil {
+			log.Warnf("failed to update sectors list: %s", err)
+		} else {
+			log.Debugf("updated sectors list")
+		}
+		return err
+	}
+
+	// Update the list immediately
+	_ = update()
+
+	// Update the list on every tick
+	for ctx.Err() == nil {
+		var err error
+		select {
+		case <-ctx.Done():
+			return
+		case onComplete := <-l.syncUpdate:
+			err = l.update(ctx)
+			onComplete(err)
+		case <-timer.C:
+			err = update()
+		}
+
+		// If there are any other threads waiting for an update, call
+		// their callbacks also
+		moreCallbacks := true
+		for moreCallbacks {
+			select {
+			case onComplete := <-l.syncUpdate:
+				onComplete(err)
+			default:
+				moreCallbacks = false
+			}
+		}
+	}
+}
+
+// Synchronously refresh the list
+func (l *SectorsList) Refresh(ctx context.Context) error {
+	done := make(chan error, 1)
+	onUpdate := func(err error) {
+		done <- err
+	}
+	select {
+	case l.syncUpdate <- onUpdate:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	select {
+	case err := <-done:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (l *SectorsList) update(ctx context.Context) error {
+	head, err := l.fullNode.ChainHead(ctx)
+	if err != nil {
+		return err
+	}
+
+	maddr, err := l.spApi.ActorAddress(ctx)
+	if err != nil {
+		return err
+	}
+
+	activeSet, err := l.fullNode.StateMinerActiveSectors(ctx, maddr, head.Key())
+	if err != nil {
+		return err
+	}
+	activeIDs := make(map[abi.SectorNumber]struct{}, len(activeSet))
+	for _, info := range activeSet {
+		activeIDs[info.SectorNumber] = struct{}{}
+	}
+
+	sset, err := l.fullNode.StateMinerSectors(ctx, maddr, nil, head.Key())
+	if err != nil {
+		return err
+	}
+	commitedIDs := make(map[abi.SectorNumber]struct{}, len(sset))
+	for _, info := range sset {
+		commitedIDs[info.SectorNumber] = struct{}{}
+	}
+
+	apiList, err := l.spApi.SectorsList(ctx)
+	if err != nil {
+		return err
+	}
+
+	sort.Slice(apiList, func(i, j int) bool {
+		return apiList[i] < apiList[j]
+	})
+
+	sectorsList := make([]SectorState, 0, len(apiList))
+	for _, sectorNum := range apiList {
+		st, err := l.spApi.SectorsStatus(ctx, sectorNum, true)
+		if err != nil {
+			return err
+		}
```

Comment:
I don't think this will work for the large SPs we want to support -- currently we have SPs with 30-50PiB of data, and we want to go up to 100PiB... this means we should think how we use these APIs that would return 1.5M up to 2M records.

I don't think we should synchronously make 2M API calls.